### ISSUE TITLE
Chapter on document ranking/retrieval

### ISSUE THREAD

**Issue by jacobdanovitch** [allenai/allennlp-guide#125]
A fairly practical topic that's maybe less straightforward than other NLP tasks. Gaining more attention lately with recent retrieval-augmented model architectures.

I might be able to contribute to [some of] this myself as well. [This repo](https://github.com/sebastian-hofstaetter/transformer-kernel-ranking) has a bunch of AllenNLP implementations already which should make the chapter a lot easier.

**Comment by matt-gardner** [comment]
That would be awesome!  If you can get something started, I'm happy to help / review / whatever else is needed.

**Comment by jacobdanovitch** [comment]
I've started this in a personal repo: [allenrank](https://github.com/jacobdanovitch/allenrank)

I have a couple of models implemented for training on the [MIMICS](https://github.com/microsoft/MIMICS) query suggestion dataset, it's a bit more lightweight than MS MARCO or any big document ranking datasets (though I can switch to those later). There's a docker-compose file in there that _should_ work if you need to run it for any reason. I'll start writing it up as a tutorial next.

I left a couple questions for you inline (you can search for `@Matt Gardner`). The first is a general pytorch/efficiency one but the second is specifically for AllenNLP: What's the best way to use metrics on a `ListField[LabelField]`? I can't use anything like AUC directly since -1 is present in the labels due to padding.

**Comment by matt-gardner** [comment]
Nice!  On your two questions:

1. What you have seems ok to me.  I don't know enough the details to know if there's anything in particular that could be fixed, but a useful thing to think about to convince yourself one way or the other is to compare our `MatrixAttention` code to its prior version.  What we used to do for `MatrixAttention` was expand the query vector out to be the size of the matrix, then compute a similarity function on top of these expanded representations.  It turns out that this is very inefficient, because of all of the intermediate computations done on large matrices.  If you can instead do the operations in pieces on smaller inputs, you can gain a bunch of efficiency.  It looks like you're taking the approach of our old version of `MatrixAttention`, but it didn't look like there was an easy way to switch to a more efficient way of doing things, because you have quite a long sequence of operations inside your `RelevanceMatcher`, and I'm not sure if it's even possible to do it with smaller intermediate computations without spending a lot more time looking at the code.  Another thing that came to mind when I saw your code was wondering whether it would be easier conceptually and implementation-wise to use a `TimeDistributed` layer in there, so the `RelevanceMatcher` only has to score a single document and query at a time, and `TimeDistributed` does the batch rolling and unrolling for you.  Not sure if that's a good idea or not, but it's something to think about.

2. For your second question about metrics, I think you're right that it'd be difficult to re-use existing code for what you want.  I was trying to decide if rolling things into the batch dimension and handling a mask appropriately might work, but I don't think these metrics allow masks.  So it seems like there are two options: (1) change the metrics to allow masks to be passed, which will let you solve your problem by rolling into the batch dimension (though this might not aggregate things how you want...), or (2) start by copying the existing metric code, and generalize it to handle lists.

Also, [these](https://github.com/jacobdanovitch/allenrank/blob/3f20c4db390d7c379350fe9511f22fb57ce27041/allenrank/models/ranker.py) are just `torch.Tensors`, not `TextFieldTensors`.

**Comment by jacobdanovitch** [comment]
Thanks for the feedback! 

1. You're right, `LegacyMatrixAttention` seems to have a similar idea to what I did. I don't think a bmm will work here instead though, since the shapes start to get a bit crazy. We have `[batch, words, dim]` for the queries and `[batch, candidates, words, dim]` for the document sets. I think it might be feasible with a careful einsum, but I don't know if that's a good idea for a tutorial; I think this way might be easier to understand (though that might be influenced by my poor understanding of einsum). What do you think? As for `TimeDistributed`, you're totally right; I'll still have to expand but that should saves the effort of expanding and shows off a super useful feature.

2. Turns out I was completely wrong, the metrics do accept masks, so no issue there. It works perfectly OOTB. The docs don't show the `__call__` signature for metrics, which might be worth adding.

Didn't have time to write this weekend but I got everything addressed above implemented as well as an NDCG metric, I'll send a draft your way towards the end of the week.

**Comment by matt-gardner** [comment]
Awesome, thanks for the update!

**Comment by jacobdanovitch** [comment]
I have a [rough draft ready](https://github.com/jacobdanovitch/allenrank/blob/master/allennlp_guide_post.md), though I haven't written up to the metrics yet. Feel free to tear it apart a bit, I wasn't sure what format was expected since nothing has been posted for part 3 yet. I assumed that they should be a bit longer than previous entries since it's an end-to-end task, so I'm basically formatting it like a blog post.

**Comment by matt-gardner** [comment]
This is really great!  I have a few (mostly minor) comments, but it'd be easier to make them on an actual PR.  Care to open a PR adding that as a chapter here, and we can iterate on it as a PR?  If you have trouble figuring out the format, don't worry too much about it, just open the PR and I can make sure things are in the right place.

**Comment by jacobdanovitch** [comment]
Sure! Should I just add the markdown file under `chapters/part3`?

**Comment by matt-gardner** [comment]
Yep, that'd be great, and I can make sure it's in the right format from there.

**Comment by jacobdanovitch** [comment]
Sorry for taking so long, kept telling myself I'd get to writing up the last section sometime this week but haven't had time. PR is opened.
